FastLED based output

netmindz opened the pull request:
Add the LED output option via FastLED rather than NeoPixelBus to compare how it performs, especially if it is lest prone to flicker/glitches

netmindz committed:
AutoPlaylist - prevent swapping playlist after silence ended when Suspended

netmindz committed:
Start adding FastLED output

netmindz committed:
Start adding FastLED output

netmindz committed:
Start adding FastLED output

netmindz committed:
Borrow FastLED.addLeds from StartModLEDs

netmindz committed:
WiP on FastLED output

softhack007 commented on wled00/bus_manager.h:
```diff
@@ -1,6 +1,8 @@
 #ifndef BusManager_h
 #define BusManager_h
 
+#include <FastLED.h>
```
seems that adding a few define's _before_ `#include <FastLED.h>` is enough to select the fastLed "clockless I2S" instead of "clockless RMT" driver
```
#define FASTLED_ESP32_I2S true               // use clockless I2S driver instead of RMT
#define I2S_DEVICE 1                         // I2S#0 needed by audioreactive
#define FASTLED_ESP32_I2S_NUM_DMA_BUFFERS 4  // recommended for solving flicker issues in combination with interrupts triggered by other code parts.
#undef FASTLED_INTERNAL // just to be sure

##include <FastLED.h>
```

The I2S implementation can drive up to 24 strips in parallel, so it might be an interesting option.

netmindz replied:
Yeah might be interesting to try as well. I think that driver is possibly older code from the same guy

netmindz committed:
Call FastLED.show once for all buses

netmindz committed:
Fix saving of pin config for FastLED

netmindz committed:
Multiple pin FastLED

softhack007 committed:
Merge branch 'mdev' into pr/134

softhack007 committed:
fastLED integration improvements

* added getPixelColor()
* adjusted IS_DIGITAL macro
* disabled FastLED color correction (already done by WLED)
* added support for RGB color order (change "#if 1" to "#if 0" to activate)
* added sanity checks for pixels out of range
* made brightness limiter happy

softhack007 committed:
move fastled include into busmanager.cpp

to solve a collision with with FX.h
which includes both  fastled.h and busmanager.h.

softhack007 committed:
fix for preview when using FastLED

FastLED stores pixels without brightness reduction, so we need to skip restoreColorLossy() when using fastLED drivers.

softhack007 commented:
Lessons learned today: FastLED clockless RMT is not compatible with NeoPixelBus - You can either have LEDs driven by one or the other, but you cannot use both drivers at the same time.

It looks like NeoPixelBus stalls waiting for RMT Interrupts - even on RMT units that are not assigned to NPB.

softhack007 committed:
make brightness limiter work with fastLED

leds[] are "full brightness", so ABL was freakin' out about high current.
